src/System.Management.Automation/engine/runtime/Binding/Binders.cs

```diff
@@ -4258,13 +4265,23 @@ private DynamicMetaObject GetIndexMultiDimensionArray(DynamicMetaObject target,
                     target.CombineRestrictions(indexes));
             }
 
-            if (getterParams.Length == 1 && getterParams[0].ParameterType == typeof(int) && CanIndexFromEndWithNegativeIndex(target))
+            if (CanIndexFromEndWithNegativeIndex(target, getter, getterParams))
             {
-                // PowerShell supports negative indexing for some types (specifically, types implementing IList or IList<T>).
+                // PowerShell supports negative indexing for types that meet the following criteria:
+                //      - Indexer method accepts one parameter that is typed as int
+                //      - The int parameter is not a type argument from a constructed generic type
+                //        (this is to exclude indexers for types that could use a negative index as
+                //        a valid key like System.Linq.ILookup)
+                //      - Declares a "Count" or "Length" property
+                //      - Does not inherit from IDictionary<> as that is handled earlier in the binder
```

Comment:
Great comments! I think they should be moved to the method `CanIndexFromEndWithNegativeIndex`  😄